lib/remote/configstageshandler.cpp

```diff
@@ -97,6 +100,11 @@ void ConfigStagesHandler::HandlePost(
 {
 	namespace http = boost::beast::http;
 
+	if (m_RunningPackageUpdates) {
+		return HttpUtility::SendJsonError(response, params, 409,
+			"Conflicting request, there is already an ongoing package update in progress. Please try it again later.");
+	}
```

Comment:
This still isn't synchronized properly. Two requests could still pass this test at the same time.

Reply:
I was thinking that since all we do here is just reading (that is, not that one thread reads and while another writes), we don't need a mutex.

Reply:
The main goal here isn't protecting `m_RunningPackageUpdates` but rather ensuring that no two requests are allowed to update config packages at the same time (that's also why I suggested https://github.com/Icinga/icinga2/pull/9267#discussion_r818876906, i.e. directly use a mutex instead of a boolean that needs to be protected by a mutex).

With the current code, two threads may do the `if (m_RunningPackageUpdates)` check at the same time and both conclude that it's safe to continue when it isn't, only later setting that marker flag.

Apart from that, the mutex also synchronizes between threads in the sense that it ensures that changes by one thread become visible in the other one, so without it, another thread would even be allowed to use some cached value. (You can have a look at [cppreference](https://en.cppreference.com/w/cpp/atomic/memory_order) but that page isn't an easy read.)

Reply:
What about an atomic variable?

Reply:
Would also work. Using a mutex has the advantage that you can use `std::unique_lock` which prevents forgetting to unlock on some code paths.

I haven't tried, but that situation might already exist in the code, I think these assignments can throw if the value isn't convertible to `bool` (via `double` I guess): https://github.com/Icinga/icinga2/blob/cecfca5592d8711d38b3b8b42fce77e299bf3462/lib/remote/configstageshandler.cpp#L120-L128

Reply:
> (that's also why I suggested [#9267 (comment)](https://github.com/Icinga/icinga2/pull/9267#discussion_r818876906), i.e. directly use a mutex instead of a boolean that needs to be protected by a mutex).

How do you check a mutex directly to see if it is locked/owned by this thread? When you unlock an already unlocked mutex is undefined behavior (I've read this before).

Reply:
Look up recursive mutexes.

Reply:
[You don't](https://en.cppreference.com/w/cpp/thread/mutex/unlock):
> The mutex must be locked by the current thread of execution, otherwise, the behavior is undefined.

If unlocking in another thread is required, just keeping a single mutex locked isn't an option. In this case, `std::atomic<bool>` is probably the best option.

Reply:
Yep.

Reply:
> I haven't tried, but that situation might already exist in the code, I think these assignments can throw if the value isn't convertible to `bool` (via `double` I guess):

```
[2022-03-07 15:30:32 +0100] warning/HttpServerConnection: Error while processing HTTP request: Can't convert 'false' to a floating point number.
```

After this, `m_RunningPackageUpdates` stays true forever!

Reply:
To ensure proper flag reset, pack it into a `Shared<Defer>` and pass the (intrusive) pointer along with the actual deploy operation.